### DISCUSSION TITLE
同期処理をサポートしてほしい

### DISCUSSION THREAD

**Discussion by koichimurakami** [mbc-net/mbc-cqrs-serverless#49] [Ideas]
CQRSで処理する際は非同期処理を基本としていますが、同期処理をしたい場面もあると思います。
また、当初同期処理で実装していたが、非同期処理に変更したい場面もあるかもしれないです。
同期処理と非同期処理ではレスポントも異なると思うので完全にフラグだけで変更出来ないと思いますが、関数の引数を同一にするなど工夫して開発者の負担を減らせるようにしたいため、このフレームワークで同期処理をサポートしてほしいです。

**Comment by koichimurakami** [comment]
本件ですが、
例えばTODOを登録する場合以下のようにFunction名にAsyncを付けるか付けないかで同期・非同期を分けるとか、Function名は同じだけどoptionsに `async: true` を付けた場合は非同期で、`async: false` を付けたら非同期にする等をイメージします。

## 非同期処理
todo = {}
service.publishAsync(todo, options) 処理内容: (Commandテーブルに書き込み→DynamoDB Strem→Lambda→StepFunction→Dataテーブルに書き込み→RDBに書き込み)

## 同期処理
todo = {}
service.publish(todo, options) 処理内容: (Dataテーブルに書き込み→RDBに書き込み)

**Reply by baquan1708** [reply]
@koichimurakami

Could you please review the change for this feature?

Bellow image is the modification description for the `publish` method. I intend to:
- add new function `publishAsync`: the logic is same as old `publish` method. When this method is called, we insert data into `command` table &#8594; DynamoDb Stream &#8594; Lamda &#8594; Step function &#8594; insert data into `history`, `data` DDB table and RDS
- add new function `publishSync`: this method support `sync` processing. When this method is called, we insert data into `data` DDB table and RDS.
![image](https://github.com/user-attachments/assets/9cdddfee-ed9f-43bb-b8c9-ed6e42af07a5)

Beside the `publish` method, we also need to reflect `sync` processing to `publishPartialUpdate` method. 
![image](https://github.com/user-attachments/assets/120450f8-af73-47c6-9798-03b645e004f9)

**Note**: With this approach, if you choose to process the command synchronously, you cannot later process it (with the same pk,sk) asynchronously and vice versa

**Reply by koichimurakami** [reply]
@baquan1708 
Could you tell me why you thought that publishPartialUpdate should not be separated?

**Reply by baquan1708** [reply]
@koichimurakami 
I'm sorry. I made a typo. Actually, `publishPartialUpdate` is split into two methods for `sync` and `async` processing: `publishPartialUpdateSync` and `publishPartialUpdateAsync`.
![image](https://github.com/user-attachments/assets/c081359e-daad-42ca-b5e8-8f24a6c997cd)

**Reply by koichimurakami** [reply]
@baquan1708 
I think it's good.
Could you tell me those interfaces? parameters, return type and so on.

**Reply by baquan1708** [reply]
Below is the interface reflecting these changes:
```ts
export interface ICommandService {
  /**
   * Publishes a command and processes it synchronously.
   *
   * @param {CommandInputModel} input - The complete command data to be published.
   * @param {ICommandOptions} opts - Options for command execution, including context and additional parameters.
   * @returns {Promise<CommandModel>} The published command model.
   * @throws {BadRequestException} If the item does not exist or the input version is invalid.
   */
  publishSync(
    input: CommandInputModel,
    opts: ICommandOptions,
  ): Promise<CommandModel>

  /**
   * Publishes a command and processes it asynchronously.
   *
   * @param {CommandInputModel} input - The complete command data to be published.
   * @param {ICommandOptions} opts - Options for command execution, including context and additional parameters.
   * @returns {Promise<CommandModel|null>} The published command model, or `null` if the command was not updated (not dirty).
   * @throws {BadRequestException} If the input version is invalid.
   */
  publishAsync(
    input: CommandInputModel,
    opts: ICommandOptions,
  ): Promise<CommandModel | null>

  /**
   * Publishes a partial command and processes it synchronously.
   *
   * @param {CommandPartialInputModel} input - The partial data used to update the command item.
   * @param {ICommandOptions} opts - Options for command execution, including context and additional parameters.
   * @returns {Promise<CommandModel>} The updated command model after the publish operation.
   * @throws {BadRequestException} If the item does not exist for the provided keys.
   */
  publishPartialUpdateSync(
    input: CommandPartialInputModel,
    opts: ICommandOptions,
  ): Promise<CommandModel>

  /**
   * Publishes a partial command and processes it asynchronously.
   *
   * @param {CommandPartialInputModel} input - The partial data used to update the command item.
   * @param {ICommandOptions} opts - Options for command execution, including context and additional parameters.
   * @returns {Promise<CommandModel>} The updated command model after the publish operation.
   * @throws {BadRequestException} If the item does not exist for the provided keys.
   */
  publishPartialUpdateAsync(
    input: CommandPartialInputModel,
    opts: ICommandOptions,
  ): Promise<CommandModel>
}
```

Other interfaces, such as `CommandInputModel`, `CommandModel`, `CommandPartialInputModel`, and `ICommandOptions`, remain unchanged.

**Reply by koichimurakami** [reply]
Thank you.
Will `publish` and `publishPartialUpdate` be deprecated in future?

**Reply by baquan1708** [reply]
Yes, they will. We mark there function as `deprecated` and refer to new function to ensure developers can easily transition to the updated implementation.

For example: 
```ts
/**
 * @deprecated Use {@link publishAsync} instead.
 * This function is outdated and will be removed in the next major release.
 */
function publish(
    input: CommandInputModel,
    opts: ICommandOptions,
  ): Promise<CommandModel | null> { ... }
```

**Reply by koichimurakami** [reply]
@baquan1708 
Thank you.

I think it's better to use English words instead of abbreviations for parameter names.  Therefore, could you please change to?

**Reply by baquan1708** [reply]
Yes, I can. I will change `opts` to `options`. Here is the revised interface:

```ts
export interface ICommandService {
  /**
   * Publishes a command and processes it synchronously.
   *
   * @param {CommandInputModel} input - The complete command data to be published.
   * @param {ICommandOptions} options - Options for command execution, including context and additional parameters.
   * @returns {Promise<CommandModel>} The published command model.
   * @throws {BadRequestException} If the item does not exist or the input version is invalid.
   */
  publishSync(
    input: CommandInputModel,
    options: ICommandOptions,
  ): Promise<CommandModel>

  /**
   * Publishes a command and processes it asynchronously.
   *
   * @param {CommandInputModel} input - The complete command data to be published.
   * @param {ICommandOptions} options - Options for command execution, including context and additional parameters.
   * @returns {Promise<CommandModel|null>} The published command model, or `null` if the command was not updated (not dirty).
   * @throws {BadRequestException} If the input version is invalid.
   */
  publishAsync(
    input: CommandInputModel,
    options: ICommandOptions,
  ): Promise<CommandModel | null>

  /**
   * Publishes a partial command and processes it synchronously.
   *
   * @param {CommandPartialInputModel} input - The partial data used to update the command item.
   * @param {ICommandOptions} options - Options for command execution, including context and additional parameters.
   * @returns {Promise<CommandModel>} The updated command model after the publish operation.
   * @throws {BadRequestException} If the item does not exist for the provided keys.
   */
  publishPartialUpdateSync(
    input: CommandPartialInputModel,
    options: ICommandOptions,
  ): Promise<CommandModel>

  /**
   * Publishes a partial command and processes it asynchronously.
   *
   * @param {CommandPartialInputModel} input - The partial data used to update the command item.
   * @param {ICommandOptions} options - Options for command execution, including context and additional parameters.
   * @returns {Promise<CommandModel>} The updated command model after the publish operation.
   * @throws {BadRequestException} If the item does not exist for the provided keys.
   */
  publishPartialUpdateAsync(
    input: CommandPartialInputModel,
    options: ICommandOptions,
  ): Promise<CommandModel>
}
```

**Reply by koichimurakami** [reply]
Thank you.
Could you proceed with the implementation of this function?

**Reply by baquan1708** [reply]
Certainly! I'll proceed with implementing this function right away.

**Reply by baquan1708** [reply]
@koichimurakami cc: @Tuan-Anh-Nguyen-pionero 

I've just created PR: https://github.com/mbc-net/mbc-cqrs-serverless/pull/74. When you have a moment, could you please review it?

**Comment by koichimurakami** [comment]
Release Info in Japanese: https://www.mbc-net.com/update-mbc-cqrs-%e3%82%b5%e3%83%bc%e3%83%90%e3%83%bc%e3%83%ac%e3%82%b9-%e3%83%95%e3%83%ac%e3%83%bc%e3%83%a0%e3%83%af%e3%83%bc%e3%82%af-ver-0-1-35-beta-core%e3%81%ab%e5%90%8c%e6%9c%9f%e5%87%a6%e7%90%86/
